File: BHAVRP_Java/BHAVRP2.0/src/cujae/inf/ic/om/controller/utils/Tools.java
package cujae.inf.ic.om.controller.utils;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Random;

import cujae.inf.ic.om.problem.input.*;

import cujae.inf.ic.om.matrix.NumericMatrix;

public abstract class Tools {
	
	public static void randomOrdenate(){
		Random random = new Random();
		int posRandom = -1;
		ArrayList<Depot> orderedDepots = new ArrayList<Depot>();
		ArrayList<Depot> copyDepots = new ArrayList<Depot>(Problem.getProblem().getDepots());
		
		while(!Problem.getProblem().getDepots().isEmpty())
		{
			posRandom = random.nextInt(Problem.getProblem().getDepots().size());
			orderedDepots.add(Problem.getProblem().getDepots().remove(posRandom));
		}
		
		for(int i = 0; i < orderedDepots.size(); i++)
			Problem.getProblem().getDepots().add(orderedDepots.get(i));
		
		NumericMatrix copyCostMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
		int currentDepots = 0;
		int totalDepots = Problem.getProblem().getTotalDepots();
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteCol(totalCustomers);			
			currentDepots++;
		}
			
		int posDepot;
		
		for(int j = 0; j < copyDepots.size(); j++)
		{
			posDepot = Problem.getProblem().findPosDepot(copyDepots, Problem.getProblem().getDepots().get(j).getIDDepot());
			copyCostMatrix.addCol((totalCustomers + j), Problem.getProblem().getCostMatrix().getCol(totalCustomers + posDepot));
		}	
		
		currentDepots = 0;
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteRow(totalCustomers);
			currentDepots++;
		}
		
		for(int k = 0; k < copyDepots.size(); k++)
		{
			posDepot = Problem.getProblem().findPosDepot(copyDepots, Problem.getProblem().getDepots().get(k).getIDDepot());
			copyCostMatrix.addRow((totalCustomers + k), Problem.getProblem().getCostMatrix().getRow(totalCustomers + posDepot));
		}	
		
		Problem.getProblem().setCostMatrix(copyCostMatrix);
	}
	
	public static void descendentOrdenate(){
		Depot depot = new Depot();
		ArrayList<Depot> copyDepots = new ArrayList<Depot>(Problem.getProblem().getDepots());
		int totalDepots = Problem.getProblem().getTotalDepots();
		
		for(int i = 0; i < (totalDepots - 1); i++)
		{
			for(int j = 0; j < (totalDepots - i - 1); j++)
			{
				if(Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepots().get(j + 1)) > Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepots().get(j)))
				{
					depot = Problem.getProblem().getDepots().get(j + 1);
					Problem.getProblem().getDepots().set((j + 1), Problem.getProblem().getDepots().get(j));
					Problem.getProblem().getDepots().set(j, depot);
				}
			}
		}
		
		NumericMatrix copyCostMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
		int currentDepots = 0;
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteCol(totalCustomers);
			currentDepots++;
		}
			
		int posDepot;
		
		for(int j = 0; j < copyDepots.size(); j++)
		{
			posDepot = Problem.getProblem().findPosDepot(copyDepots, Problem.getProblem().getDepots().get(j).getIDDepot());
			copyCostMatrix.addCol((totalCustomers + j), Problem.getProblem().getCostMatrix().getCol(totalCustomers + posDepot));
		}	
		
		currentDepots = 0;
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteRow(totalCustomers);
			currentDepots++;
		}
		
		for(int k = 0; k < copyDepots.size(); k++)
		{
			posDepot = Problem.getProblem().findPosDepot(copyDepots, Problem.getProblem().getDepots().get(k).getIDDepot());
			copyCostMatrix.addRow((totalCustomers + k), Problem.getProblem().getCostMatrix().getRow(totalCustomers + posDepot));
		}	
		
		Problem.getProblem().setCostMatrix(copyCostMatrix);
	}

	public static void ascendentOrdenate(){
		Depot depot = new Depot();
		ArrayList<Depot> copyDepots = new ArrayList<Depot>(Problem.getProblem().getDepots());
		int totalDepots = Problem.getProblem().getTotalDepots();
		
		for(int i = 0; i < (totalDepots - 1); i++)
		{
			for(int j = 0; j < (totalDepots - i - 1); j++)
			{
				if(Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepots().get(j + 1)) < Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepots().get(j)))
				{
					depot = Problem.getProblem().getDepots().get(j + 1);
					Problem.getProblem().getDepots().set((j + 1), Problem.getProblem().getDepots().get(j));
					Problem.getProblem().getDepots().set(j, depot);
				}
			}
		}
		
		NumericMatrix copyCostMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
		int currentDepots = 0;
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteCol(totalCustomers);
			currentDepots++;
		}
			
		int posDepot;
		
		for(int j = 0; j < copyDepots.size(); j++)
		{
			posDepot = Problem.getProblem().findPosDepot(copyDepots, Problem.getProblem().getDepots().get(j).getIDDepot());
			copyCostMatrix.addCol((totalCustomers + j), Problem.getProblem().getCostMatrix().getCol(totalCustomers + posDepot));
		}	
		
		currentDepots = 0;
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteRow(totalCustomers);
			currentDepots++;
		}
		
		for(int k = 0; k < copyDepots.size(); k++)
		{
			posDepot = Problem.getProblem().findPosDepot(copyDepots, Problem.getProblem().getDepots().get(k).getIDDepot());
			copyCostMatrix.addRow((totalCustomers + k), Problem.getProblem().getCostMatrix().getRow(totalCustomers + posDepot));
		}	
		
		Problem.getProblem().setCostMatrix(copyCostMatrix);
	}

	public static void randomOrdenate(ArrayList<Customer> customers){
		Random random = new Random();
		int posRandom = -1;
		ArrayList<Customer> orderedCustomers = new ArrayList<Customer>();
		
		while(!customers.isEmpty())
		{
			posRandom = random.nextInt(customers.size());
			orderedCustomers.add(customers.remove(posRandom));
		}
		
		for(int i = 0; i < orderedCustomers.size(); i++)
			customers.add(orderedCustomers.get(i));
		
		/* no necesito actualizar la matriz*/
	}

	public static void descendentOrdenate(ArrayList<Customer> customers){
		Customer customer = new Customer();
		int totalCustomers = customers.size();
		
		for(int i = 0; i < (totalCustomers - 1); i++)
		{
			for(int j = 0; j < (totalCustomers - i - 1); j++)
			{
				if(Problem.getProblem().getRequestByIDCustomer(customers.get(j + 1).getIDCustomer()) > Problem.getProblem().getRequestByIDCustomer(customers.get(j).getIDCustomer()))
				{
					customer = customers.get(j + 1);
					customers.set((j + 1), customers.get(j));
					customers.set(j, customer);
					
					/* no actualizo infoproblem*/
				}
			}
		}
	}

	public static void ascendentOrdenate(ArrayList<Customer> customers){
		Customer customer = new Customer();
		int totalCustomers = customers.size();
		
		for(int i = 0; i < (totalCustomers - 1); i++)
		{
			for(int j = 0; j < (totalCustomers - i - 1); j++)
			{
				if(Problem.getProblem().getRequestByIDCustomer(customers.get(j + 1).getIDCustomer()) < Problem.getProblem().getRequestByIDCustomer(customers.get(j).getIDCustomer()))
				{
					customer = customers.get(j + 1);
					customers.set(j+1, customers.get(j));
					customers.set(j, customer);
				}
			}
		}
	}
	
	public static double truncateDouble(double number, int decimalPlace) {
        double numberRound;
        BigDecimal bigDec = new BigDecimal(number);
        bigDec = bigDec.setScale(decimalPlace, BigDecimal.ROUND_DOWN);
        numberRound = bigDec.doubleValue();
        
        return numberRound;
    }
}

File: BHAVRP_Python/cujae/inf/citi/om/controller/tools/tools.py
import random
import numpy as np
from typing import List
from ...assignment.assignment import Assignment
from ...service.distance_type import DistanceType
from ...problem.input.problem import Problem
from ...problem.input.customer import Customer

class Tools:
        
    @staticmethod
    def random_ordenate():
        random.shuffle(Problem.get_problem().get_depots())
        distance_type: DistanceType = Assignment.distance_type 
        
        copy_cost_matrix = Assignment.initialize_cost_matrix(Problem.get_problem().get_customers(), copy_depots, distance_type)
        total_depots = Problem.get_problem().get_total_depots()
        total_customers = Problem.get_problem().get_total_customers()
        
        for _ in range(total_depots):
            copy_cost_matrix = np.delete(copy_cost_matrix, total_customers, axis=1)
        
        copy_depots = list(Problem.get_problem().get_depots())
        
        for j in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[j].id_depot)
            col_index = total_customers + pos_depot 
            while col_index >= copy_cost_matrix.shape[1]:
                new_col = np.zeros((copy_cost_matrix.shape[0], 1))
                copy_cost_matrix = np.hstack((copy_cost_matrix, new_col))
            new_col = Problem.get_problem().cost_matrix[:, col_index].reshape(-1, 1)
            copy_cost_matrix = np.hstack((copy_cost_matrix, new_col))

        for _ in range(total_depots):
            copy_cost_matrix = np.delete(copy_cost_matrix, total_customers, axis=0)
        
        for k in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[k].id_depot)
            row_index = total_customers + pos_depot 
            if row_index < copy_cost_matrix.shape[0]:
                new_row = copy_cost_matrix[row_index, :].reshape(1, -1) 
                copy_cost_matrix = np.vstack((copy_cost_matrix, new_row))
            else:
                new_row = copy_cost_matrix[-1, :].reshape(1, -1)
                copy_cost_matrix = np.vstack((copy_cost_matrix, new_row)) 

        Problem.get_problem().set_cost_matrix(copy_cost_matrix)
        
    @staticmethod
    def descendent_ordenate():
        copy_depots = list(Problem.get_problem().get_depots())
        total_depots = len(copy_depots)
        distance_type: DistanceType = Assignment.distance_type    
            
        for i in range(total_depots - 1):
            for j in range(total_depots - i - 1):
                if Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j + 1]) > Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j]):
                    Problem.get_problem().depots[j], Problem.get_problem().depots[j + 1] = Problem.get_problem().depots[j + 1], Problem.get_problem().depots[j]
            
        copy_cost_matrix = Assignment.initialize_cost_matrix(Problem.get_problem().get_customers(), copy_depots, distance_type)
        total_customers = Problem.get_problem().get_total_customers()
            
        for _ in range(total_depots):
            if total_customers < copy_cost_matrix.shape[1]:
                copy_cost_matrix = np.delete(copy_cost_matrix, total_customers - 1, axis=1)
            
        for j in range(total_depots):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().get_depots()[j].get_id_depot())
            col_index = total_customers + pos_depot - 1
            while col_index >= copy_cost_matrix.shape[1]:
                new_col = np.zeros((copy_cost_matrix.shape[0], 1))  # Nueva columna con ceros
                copy_cost_matrix = np.hstack((copy_cost_matrix, new_col))
            new_col = copy_cost_matrix[:, col_index].reshape(-1, 1)
            copy_cost_matrix = np.hstack((copy_cost_matrix, new_col))

        for _ in range(total_depots):
            if total_customers < copy_cost_matrix.shape[0]:
                copy_cost_matrix = np.delete(copy_cost_matrix, total_customers - 1, axis=0)
            
        for k in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().get_depots()[k].get_id_depot())
            row_index = total_customers + pos_depot - 1
            if row_index < copy_cost_matrix.shape[0]:
                new_row = copy_cost_matrix[row_index, :].reshape(1, -1)
                copy_cost_matrix = np.vstack((copy_cost_matrix, new_row))
            else:
                new_row = copy_cost_matrix[-1, :].reshape(1, -1)
                copy_cost_matrix = np.vstack((copy_cost_matrix, new_row))
                
        Problem.get_problem().set_cost_matrix(copy_cost_matrix)

    @staticmethod
    def ascendent_ordenate():
        copy_depots = list(Problem.get_problem().get_depots())
        total_depots = len(copy_depots)
        distance_type: DistanceType = Assignment.distance_type
            
        for i in range(total_depots - 1):
            for j in range(total_depots - i - 1):
                if Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j + 1]) < Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j]):
                    Problem.get_problem().depots[j], Problem.get_problem().depots[j + 1] = Problem.get_problem().depots[j + 1], Problem.get_problem().depots[j]
            
        copy_cost_matrix = Assignment.initialize_cost_matrix(Problem.get_problem().get_customers(), copy_depots, distance_type)
        total_customers = Problem.get_problem().get_total_customers()
            
        for _ in range(total_depots):
            copy_cost_matrix = np.delete(copy_cost_matrix, total_customers, axis=1)
            
        for j in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().get_depots()[j].get_id_depot())
            col_index = total_customers + pos_depot 
            while col_index >= copy_cost_matrix.shape[1]:
                new_col = np.zeros((copy_cost_matrix.shape[0], 1))
                copy_cost_matrix = np.hstack((copy_cost_matrix, new_col))
            new_col = Problem.get_problem().cost_matrix[:, col_index].reshape(-1, 1)
            copy_cost_matrix = np.hstack((copy_cost_matrix, new_col))
            
        for _ in range(total_depots):
            copy_cost_matrix = np.delete(copy_cost_matrix, total_customers, axis=0)
            
        for k in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().get_depots()[k].get_id_depot())
            row_index = total_customers + pos_depot
            if row_index < copy_cost_matrix.shape[0]:
                new_row = copy_cost_matrix[row_index, :].reshape(1, -1)
                copy_cost_matrix = np.vstack((copy_cost_matrix, new_row))
            else:
                new_row = copy_cost_matrix[-1, :].reshape(1, -1)
                copy_cost_matrix = np.vstack((copy_cost_matrix, new_row))
                
        Problem.get_problem().set_cost_matrix(copy_cost_matrix)

    @staticmethod
    def random_ordenate_customers(customers: List[Customer]):
        random.shuffle(customers)

    @staticmethod
    def descendent_ordenate_customers(customers: List[Customer]):
        customers.sort(key=lambda c: Problem.get_problem().get_request_by_id_customer(c.id_customer), reverse=True)

    @staticmethod
    def ascendent_ordenate_customers(customers: List[Customer]):
        customers.sort(key=lambda c: Problem.get_problem().get_request_by_id_customer(c.id_customer))

    @staticmethod
    def truncate_double(number: float, decimal_place: int) -> float:
        return round(number, decimal_place)